### PR TITLE
Add AdoptOpenJDK

### PR DESCRIPTION
**Pull request by ta2gch**
I added Formulae for AdoptOpenJDK.
This issue relates #12704 .

### CODE REVIEW

**Comment by iMichka** [comment]
Hi.

Can you remove the `update-adoptopenjdk.rb` file? We do not keep that kind of code in this repository, only formulae.

I have two other questions (because I did not follow the whole AdoptOpenJDK discussion:
- could this be added to the mac repo first?
- Why are there no jdk 9 and 10 formulae?

cc @sjackman

**Comment by sjackman** [comment]
> could this be added to the mac repo first?

There's a cask for it: https://github.com/Homebrew/homebrew-cask/blob/master/Casks/adoptopenjdk.rb
There are no JDK formulae in Homebrew/core.

> Why are there no jdk 9 and 10 formulae?

JDK 9 broke compatibility in a pretty big way (though I'm not sure of the details) with JDK 8. So many tools require JDK ≤ 8, but not many specifically required JDK 9 or 10. AdoptOpenJDK packages only JDK 8, 11, and 12. See https://adoptopenjdk.net

**Comment by ta2gch** [comment]
@sjackman Thank you for answering, That's true AdoptOpenJDK is available int the homebrew-cask, because it is distributed with `.pkg` style. And also, I added the versions 8 and 11 because those versions are marked LTS. however I can add 12 soon if you want. I'll remove the update script.

> JDK 9 broke compatibility in a pretty big way (though I'm not sure of the details) with JDK 8.

By the way, I got the information that scientific software might not run on the AdoptOpenJDK because that JDK doesn't include JavaFX. Oracle's and some others also doesnt have that.
Currently Liberica JDK and Zulu JDK, and Amazon Corretto 8  keeps compatibility backward versions that means these JDK have JavaFX implementation. I suggest you to add one of them.

**Comment by sjackman** [comment]
Following the principle of YAGNI (you aren't going to need it), I suggest not worrying about JavaFX until a user asks for it.

**Comment by ta2gch** [comment]
@sjackman I see.
@iMichka I deleted it.

**Comment by ta2gch** [comment]
@sjackman I solved my worry 😄 because Liberica JDK is going to make its tap for Homebrew/Linuxbrew.
https://github.com/bell-sw/Liberica/issues/5

**Comment by sjackman** [comment]
Please address these style issues:
```
  * C: 1: col 5: Layout/InitialIndentation: Indentation of first line in file detected.
  * C: 6: col 7: FormulaAudit/Options: Formulae should not have an `option`
  * C: 8: col 1: Layout/TrailingWhitespace: Trailing whitespace detected.
  * C: 11: col 9: Layout/IndentationWidth: Use 2 (not 4) spaces for indentation.
  * C: 12: col 189: Metrics/LineLength: Line is too long. [223/170]
  * C: 17: col 1: Layout/TrailingWhitespace: Trailing whitespace detected.
```
https://circleci.com/gh/Homebrew/linuxbrew-core/27876

**Comment by ta2gch** [comment]
I hope all warning is solved ... In the fact there is no warning in local environment.

**Comment by sjackman** [comment]
`brew audit --new-formula --strict adoptopenjdk` and `brew style adoptopenjdk` should give you these warnings.

**Comment by sjackman** [comment]
`brew audit --new-formula adoptopenjdk` and `brew style --strict adoptopenjdk` should give you these warnings.

```
  * C: 11: col 3: FormulaAudit/Options: Formulae should not have an `option`
```
This warning remains. Please remove the `with-openj9` option.

**Comment by sjackman** [comment]
```
==> Downloading https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.3%2B7/OpenJDK11U-jdk_x64_linux_hotspot_11.0.3_7.tar.gz
Error: An exception occurred within a child process:
  ChecksumMismatchError: SHA256 mismatch
Expected: 39dad3e7d94fd8d313d24563677316cb72c0490af50de4f852bc5217e58c098e
  Actual: 23cded2b43261016f0f246c85c8948d4a9b7f2d44988f75dad69723a7a526094
```
https://circleci.com/gh/Homebrew/linuxbrew-core/27877

**Comment by ta2gch** [comment]
Okay, thank you for informing. I see. Could I add **with-openj9** without warnings?

**Comment by sjackman** [comment]
Linuxbrew/core does not make use of options. Someone who wanted to use openj9 could if they liked use the formula from https://github.com/AdoptOpenJDK/homebrew-openjdk.

**Comment by ta2gch** [comment]
I see. So you said I should separate the fomula each options if someone want it, don't you?
Actually, there are no formula in that repository because AdoptOpenJDK/homebrew-openjdk provides Casks only. Anyway, I got it I remove that option and and try to fix all warning.

**Comment by ta2gch** [comment]
I'm sorry. I have to leave here because I'm be busy since today. I resume this work after a few weeks.

**Comment by sjackman** [comment]
```
  * C: 2: col 9: FormulaAuditStrict/Desc: Description shouldn't start with the formula name
```
https://circleci.com/gh/Homebrew/linuxbrew-core/27893

The other warnings we usually ignore for JDK formulae.

**Comment by ta2gch** [comment]
I got it, I fix now.

**Comment by sjackman** [comment]
```
  * Formula has other versions so create a versioned alias:
      cd /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/adoptopenjdk.rb adoptopenjdk@3
```
The default version of `openjdk` is version 8. I would prefer the default version of `adoptopenjdk` to also be version 8. Please rename `adoptopenjdk.rb` to `adoptopenjdk@11.rb` and rename `adoptopenjdk@8.rb` to `adoptopenjdk.rb` and create this symlink `ln -s ../Formula/adoptopenjdk.rb adoptopenjdk@8`.

**Comment by sjackman** [comment]
```
  * A top-level "man" directory was found
    Homebrew requires that man pages live under share.
    This can often be fixed by passing "--mandir=#{man}" to configure.
```
This issue can be fixed with
```ruby
share.install prefix/"man"
```

File: Formula/adoptopenjdk.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class AdoptopenjdkAT8 < Formula
```
Comment (sjackman) [inline Formula/adoptopenjdk.rb]:
```suggestion
class Adoptopenjdk < Formula
```

File: Formula/adoptopenjdk.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  desc "Prebuilt binaries produced from OpenJDK class libraries"
+  homepage "https://adoptopenjdk.net/"
+  url "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u212-b03/OpenJDK8U-jdk_x64_linux_hotspot_8u212b03.tar.gz"
+  sha256 "dd28d6d2cde2b931caf94ac2422a2ad082ea62f0beee3bf7057317c53093de93"
+  depends_on :linux
```
Comment (sjackman) [inline Formula/adoptopenjdk.rb]:
```suggestion

  depends_on :linux
```

Comment (sjackman) [reply Formula/adoptopenjdk.rb]: Include a blank line between `sha256` and `depends_on :linux`

File: Formula/adoptopenjdk.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  def install
+    prefix.install Dir["*"]
+    share.mkdir
```
Comment (sjackman) [inline Formula/adoptopenjdk.rb]:
```suggestion
```
This line is not needed.

File: Formula/adoptopenjdk@11.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+class AdoptopenjdkAT11 < Formula
+  desc "Prebuilt binaries produced from OpenJDK class libraries"
+  homepage "https://adoptopenjdk.net/"
+  url "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.3%2B7/OpenJDK11U-jdk_x64_linux_hotspot_11.0.3_7.tar.gz"
+  sha256 "23cded2b43261016f0f246c85c8948d4a9b7f2d44988f75dad69723a7a526094"
+  depends_on :linux
```
Comment (sjackman) [inline Formula/adoptopenjdk@11.rb]:
```suggestion

  depends_on :linux
```

File: Formula/adoptopenjdk@11.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  def install
+    prefix.install Dir["*"]
+    share.mkdir
```
Comment (sjackman) [inline Formula/adoptopenjdk@11.rb]:
```suggestion
```

**Comment by sjackman** [comment]
```
  * Formula has other versions so create a versioned alias:
      cd /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/adoptopenjdk.rb adoptopenjdk@03
  * Formula has invalid versioned aliases:
      adoptopenjdk@8
```
`brew` is incorrectly detecting the version of `adoptopenjdk` 8 as version `03` and the version of `adoptopenjdk@11` as version `3.7`. You'll need to add an explicit `version` statement to both of these.

**Comment by ta2gch** [comment]
@sjackman What  format is correct?
```
  * Stable: version 11.0.3_7 should not end with an underline and a number
```

**Comment by ta2gch** [comment]
```
  * Formula has other versions so create a versioned alias:
      cd /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/adoptopenjdk.rb adoptopenjdk@8u212-b03
```

**Comment by sjackman** [comment]
I suggest `8.0.212` for `8u212-b03` and `11.0.3.7` for `11.0.3_7`.
What does `java -version` report for these two tools?

**Comment by ta2gch** [comment]
```
vagrant@ubuntu-bionic:~$ jdk8u212-b03/bin/java -version
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)
vagrant@ubuntu-bionic:~$ jdk-11.0.3+7/bin/java -version
openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.3+7, mixed mode)
```

File: Formula/adoptopenjdk.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class Adoptopenjdk < Formula
+  desc "Prebuilt binaries produced from OpenJDK class libraries"
+  homepage "https://adoptopenjdk.net/"
+  url "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u212-b03/OpenJDK8U-jdk_x64_linux_hotspot_8u212b03.tar.gz"
+  version "8.0.212"
```
Comment (sjackman) [inline Formula/adoptopenjdk.rb]:
```suggestion
  version "1.8.0.212"
```
Let's use `1.8.0.212`. Sorry for the back and forth. We're almost there.

Comment (ta2gch) [reply Formula/adoptopenjdk.rb]: I got it. I change the format soon.

**Comment by iMichka** [comment]
Thanks @ta2gch !

**Comment by ta2gch** [comment]
I glad to use linuxbrew to install adoptopenjdk!  Thanks too :)

**Pull request closed without merging**